nengo/transforms.py

```diff
@@ -0,0 +1,233 @@
+
+        if isinstance(input_shape, ConvShape):
+            assert input_shape.channels_last == channels_last
+        else:
+            input_shape = ConvShape(input_shape, channels_last=channels_last)
+
+        self.n_filters = n_filters
+        self.input_shape = input_shape
+        self.kernel_size = kernel_size
+        self.strides = strides
+        self.padding = padding
+        self.channels_last = channels_last
+        self.kernel = kernel
+        self.dimensions = len(kernel_size)
+
+        if len(kernel_size) != input_shape.dimensions:
+            raise ValidationError(
+                "Kernel dimensions (%d) do not match input dimensions (%d)" % (
+                    len(kernel_size), input_shape.dimensions), "kernel_size")
+        if len(strides) != input_shape.dimensions:
+            raise ValidationError(
+                "Stride dimensions (%d) do not match input dimensions (%d)" % (
+                    len(strides), input_shape.dimensions), "strides")
+        if not isinstance(kernel, Distribution):
+            if kernel.shape != self.kernel_shape:
+                raise ValidationError(
+                    "Kernel shape %s does not match expected shape %s" % (
+                        kernel.shape, self.kernel_shape), "kernel")
+
+        # compute output shape
+        output_shape = np.array(input_shape.spatial_shape, dtype=np.float64)
+        if self.padding == "valid":
+            output_shape -= kernel_size
+            output_shape += 1
+        output_shape /= strides
+        output_shape = tuple(np.ceil(output_shape).astype(np.int64))
+        output_shape = (output_shape + (n_filters,) if channels_last else
+                        (n_filters,) + output_shape)
+
+        self.output_shape = ConvShape(
+            output_shape, channels_last=channels_last)
+
+        self._paramdict = {
+            "n_filters": n_filters,
+            "input_shape": self.input_shape,
+            "kernel_size": self.kernel_size,
+            "strides": self.strides
+        }
+
+    @property
+    def kernel_shape(self):
+        return self.kernel_size + (self.input_shape.n_channels, self.n_filters)
+
+    def sample(self, rng=np.random):
+        if isinstance(self.kernel, Distribution):
+            kernel = []
+            # we sample this way so that any variancescaling distribution based
+            # on n/d is scaled appropriately
+            for _ in range(np.prod(self.kernel_size)):
+                kernel.append(self.kernel.sample(self.input_shape.n_channels,
+                                                 self.n_filters, rng=rng))
+            kernel = np.reshape(kernel, self.kernel_shape)
+        else:
+            kernel = np.array(self.kernel)
+        return kernel
+
+    @property
+    def size_in(self):
+        return self.input_shape.size
+
+    @property
+    def size_out(self):
+        return self.output_shape.size
+
+
+class ConvShape(object):
```

Comment:
This could potentially be used in the future for non-convolutional things. For example, you could have local filters which are like convolution but different at different image locations. These still deal in channels. I'm wondering if we should call it `ChannelShape` or something more generic.